feat!: Multiple Chunk Upload Support

danielbankhead opened the pull request:
Adding support for [multiple chunk uploads](https://cloud.google.com/storage/docs/performing-resumable-uploads#chunked-upload). Required a sizable refactor and dozens of new tests - potentially breaking change.

d-goog committed:
feat: Init multi chunk upload support

d-goog committed:
test: Fix existing tests by adding missing chunk stream

ddelgrosso1 commented on src/index.ts:
```diff
@@ -296,6 +307,8 @@ export class Upload extends Pumpify {
     if (typeof cfg.generation === 'number') {
       cacheKeyElements.push(`${cfg.generation}`);
     }
+    // TODO: add another element for chunk size? Any conflicts possible here?
```
The only situation I can imagine needing a second key would be if the a user started a non-chunk size upload that fails and then starts a chunk size upload before resuming the first. I don't see this is a very likely scenario.

ddelgrosso1 commented:
Just a general comment, I took a look at the newly added code in `index.ts` and it makes sense to me. It looks like as of now all the existing unit and conformance tests still work. Have you verified that you do indeed see multiple HTTP requests when a chunk size is supplied, i.e. against test bench?

danielbankhead commented:
Working on verification now, along with additional tests. Thanks for the preview review!

d-goog committed:
feat: Add support for multi-chunk upload

- https://cloud.google.com/storage/docs/performing-resumable-uploads#chunked-upload

d-goog committed:
docs: Document `chunkSize` option

d-goog committed:
refactor: Refactor, bug fixes, documentation, & clarification

d-goog committed:
refactor: Lint, bug fixes, and clarification

d-goog committed:
style: Make linter happy

d-goog committed:
refactor: Minor clean up and clarification

d-goog committed:
fix: Fix comp w/ Node 10

API `writable.writableEnded` isn't available until Node 12.9.0

d-goog committed:
test: Update tests for multi chunk support and improve existing ones

d-goog committed:
build: Remove now unused 'is-stream'

d-goog committed:
Merge branch 'main' of https://github.com/googleapis/gcs-resumable-upload into multi-upload-by-chunk-support

ddelgrosso1 commented on package.json:
```diff
@@ -56,15 +56,13 @@
     "@types/sinon": "^10.0.0",
     "c8": "^7.0.0",
     "gts": "^2.0.0",
-    "is-stream": "^2.0.0",
```
Out of curiosity, what did this do and why is it safe for removal?

danielbankhead replied:
It was used to validate if an object is a stream in tests. It’s no longer needed.

ddelgrosso1 commented on src/index.ts:
```diff
@@ -92,9 +97,7 @@ export interface UploadConfig {
-    request: <T = any>(
-      opts: GaxiosOptions
-    ) => Promise<GaxiosResponse<T>> | GaxiosPromise<T>;
+    request: (opts: GaxiosOptions) => Promise<GaxiosResponse> | GaxiosPromise;
```
Was this change because it was defined as `any` type?

danielbankhead replied:
Yep

danielbankhead replied:
Good point - I reverted this change with a generic `<T>` (without the `any`)

ddelgrosso1 commented:
Thanks for adding all of those tests.

ddelgrosso1 commented on test/test.ts:
```diff
@@ -46,6 +45,9 @@ class ConfigStore {
   }
 }
 
+const RESUMABLE_INCOMPLETE_STATUS_CODE = 308;
```
Just for consistency I noticed a few other places are checking status codes directly against a hardcoded number. It might make sense to either add them as constants or remove this and just compare directly to the code.

ddelgrosso1 replied:
Never mind I just noticed this was in the tests file.

ddelgrosso1 commented on src/index.ts:
```diff
@@ -476,138 +662,234 @@ export class Upload extends Pumpify {
-      const firstChunk = chunk.slice(0, 16).valueOf();
+      // continue uploading next chunk
+      this.continueUploading();
+    } else if (resp.status < 200 || resp.status > 299) {
```
Following on my comment from the tests file, maybe make these constants if other HTTP codes are being used as constants?

danielbankhead replied:
Makes sense - I’ll change the range checks to functions (e.g. the 2xx check) and 404 a constant

tritone commented on src/index.ts:
```diff
@@ -704,12 +988,20 @@ export class Upload extends Pumpify {
-    this.numBytesWritten = 0;
+    if (this.numBytesWritten) {
+      const message =
+        'Attempting to restart an upload after unrecoverable bytes have been written. Stopping as this could result in data loss.';
```
This error message might be confusing for end users. What's actionable for them here?

danielbankhead replied:
Good point - I'll recommend creating a new upload object.

tritone commented on src/index.ts:
```diff
@@ -476,138 +662,234 @@ export class Upload extends Pumpify {
+        // be fine for the next request.
+        this.unshiftChunkBuffer(dataToPrependForResending);
+        this.numBytesWritten -= missingBytes;
+        this.lastChunkSent = Buffer.alloc(0);
```
Is the buffer re-allocated for each chunk?

danielbankhead replied:
Yes, from upstream.
`Buffer.alloc(0)` creates a "Fast" buffer: https://github.com/nodejs/node/blob/e4aa575b05fcf560c6b77b0fa5036328a3417442/lib/buffer.js#L359-L366

tritone replied:
Okay-- just asking because I know in some languages repeated buffer allocs can be painful. Is there anyway to create one buffer and reuse it for each chunk? Or is that not really how it works in node?

danielbankhead replied:
Understandable - it doesn't really work that way in Node.js (at least for `Buffer.alloc(0)`)

tritone reviewed (commented):
Couple questions, I'm not a node expert so I'll rely on @ddelgrosso1 to give a more comprehensive review!

tritone reviewed (commented):
Couple questions, I'm not a node expert so I'll rely on @ddelgrosso1 to give a more comprehensive review!

d-goog committed:
refactor: Use constants and methods for status checks

danielbankhead commented on README.md:
```diff
@@ -294,6 +294,16 @@ The maximum number of retries to attempt. Default: 5.
+
+Enables [Multiple chunk upload](https://cloud.google.com/storage/docs/performing-resumable-uploads#chunked-upload) mode and sets each request size to this amount.
+
+The chunk size should be a multiple of 256 KiB (256 x 1024 bytes), unless it's the last chunk that completes the upload. Larger chunk sizes typically make uploads more efficient. We recommend using at least 8 MiB for the chunk size.
```
Thanks, I'll update and link to documentation - this way any changes in data/guidance can be referenced by the canonical source across the langs.

tritone replied:
Sounds good!

d-goog committed:
docs: Clarification for `chunkSize`

danielbankhead commented on src/index.ts:
```diff
@@ -362,7 +384,21 @@ export class Upload extends Pumpify {
     this.contentLength = isNaN(contentLength) ? '*' : contentLength;
 
+    // For backwards-compatibility with Node v10. The more convenient
+    // `this.upstream.writableEnded` is available after v12.9.0.
```
Thanks - I think we can remove this comment altogether; `upstreamEnded` is a bit easier to understand than using `this.upstream.writableEnded`

danielbankhead commented on src/index.ts:
```diff
@@ -362,7 +384,21 @@ export class Upload extends Pumpify {
+
     this.once('writing', () => {
+      // Now that someone is writing to this object, let's attach
+      // some duplexes. These duplexes enable this object to be
```
Yes, it's a common term in both networking and Node.js:
-  https://en.wikipedia.org/wiki/Duplex_(telecommunications)#FULL-DUPLEX
- https://nodejs.org/api/stream.html#implementing-a-duplex-stream

danielbankhead commented on src/index.ts:
```diff
@@ -377,6 +413,155 @@ export class Upload extends Pumpify {
+  }
+
+  /**
+   * Prepends data back to the upstream chunk buffer.
```
It's copying data: https://github.com/nodejs/node/blob/master/lib/buffer.js#L535-L574

d-goog committed:
docs: Clarifications in documentation

danielbankhead commented on src/index.ts:
```diff
@@ -103,6 +106,16 @@ export interface UploadConfig {
+  /**
+   * Create a separate request per chunk.
+   *
+   * Should be a multiple of 256 KiB (2^18).
```
Yep - validated. The error message (truncated): `Invalid request.  The number of bytes uploaded is required to be equal or greater than 262144, except for the final request (it's recommended to be the exact multiple of 262144)`

ddelgrosso1 commented on src/index.ts:
```diff
@@ -490,123 +515,255 @@ export class Upload extends Pumpify {
+        }
+
+        if (result.done) {
+          leadStream.push(null);
```
I didn't look closely but just want to verify that using `null` as the argument here won't cause problems?

danielbankhead replied:
`null` tells signals to consumers 'no more data will be available from this stream': https://nodejs.org/api/stream.html#readablepushchunk-encoding

ddelgrosso1 commented on src/index.ts:
```diff
@@ -377,6 +412,155 @@ export class Upload extends Pumpify {
+
+    // notify upstream we've read from the buffer so it can potentially
+    // send more data down.
+    process.nextTick(() => this.emit('readFromChunkBuffer'));
```
Why does this need to occur on `nextTick`?

danielbankhead replied:
Slight optimization - gives the opportunity for the chunk to be passed downstream before requesting more data to be immediately read

ddelgrosso1 commented:
I think overall the refactor looks good. There are a few failing tests in CI.

d-goog committed:
test: Pass status in `responseHandler` tests

danielbankhead commented:
@ddelgrosso1 thanks - just fixed some tests for the `#isSuccessfulResponse` refactor

ddelgrosso1 reviewed (approved):
Overall I don't have any major concerns, I think the code and the tests look good. I would like some other eyes on it before merging though.

bcoe commented on src/index.ts:
```diff
@@ -476,138 +661,234 @@ export class Upload extends Pumpify {
+   *
+   * @returns if the request is ok to continue as-is
+   */
+  private async ensureUploadingSameObject() {
```
Is this edge case likely to happen? and what would have happened in the previous implementation of `gcs-resumbable-upload`?

If this is an optimization, it might be nice to do this in follow up work just to simplify this PR.

danielbankhead replied:
This preserves existing behavior - I don't think this is likely to happen (https://github.com/googleapis/gcs-resumable-upload/blob/main/src/index.ts#L577-L596)

bcoe commented on test/test.ts:
```diff
@@ -467,311 +830,446 @@ describe('gcs-resumable-upload', () => {
+          assert.equal(data.byteLength, CHUNK_SIZE);
+        });
+
+        it('should prepare a valid request if the remaining data is less than `chunkSize`', async () => {
```
I appreciate how thorough these tests are 🎉

bcoe commented on src/index.ts:
```diff
@@ -476,138 +661,234 @@ export class Upload extends Pumpify {
-      let cachedFirstChunk = this.get('firstChunk');
-      const firstChunk = chunk.slice(0, 16).valueOf();
+      // continue uploading next chunk
+      this.continueUploading();
```
It looks like the logic of `continueUploading` hasn't changed much. Is the upload endpoint a constant URI, and it's just `Content-Range` that changes to support multiple chunk uploads?

danielbankhead replied:
Correct - it's mainly `Content-Range` and supporting N requests per `Upload` (required a lot of decoupling throughout the class)

bcoe commented on src/index.ts:
```diff
@@ -476,138 +661,234 @@ export class Upload extends Pumpify {
+      resp.status === RESUMABLE_INCOMPLETE_STATUS_CODE &&
+      resp.headers.range;
+
+    if (shouldContinueWithNextMultiChunkRequest) {
```
There are a few places like this, where we have almost completely different logic for chunk upload, vs., regular upload, but share a method.

It makes me wonder if the implementation would read slightly cleaner if we had, `export class Upload extends Pumpify` in `index.ts`, as it is today, but then introduced an additional class, `export class ChunkUpload extends Upload`.

It might introduce a bit of repeated logic, but I'm curious if it would make the implementation even more elegant.

tritone replied:
+1, I think separating out the logic more between the two upload types would definitely make the code easier to read.

danielbankhead replied:
Good point - I figured after the refactor to support multiple requests most of the logic would be the same with the exception of 1 parameter (`upstreamIterator` in `startUploading`) and 3 `if` statements (including this one).

I was hoping `shouldContinueWithNextMultiChunkRequest` would make it easier to understand what's happening, but seemed to have backfired.

bcoe replied:
As I said, not blocking for me, but if you have the time to experiment with the refactor, I'd be curoius what it looks like.

bcoe reviewed (changes requested):
A few thoughts:

1. I would be tempted to call this a breaking change, since you rewrite a lot of the original implementation, along with adding the chunk upload support, I think a major bump will help protect you from any regressions, because you can make the update to `@google-cloud/storage` explicit, and test throughly.
2. When you do add this to `@google-cloud/storage`, I would add a few integration tests -- I think you could just generate junk data, and assert that both types of uploading work effectively.
3. I left a note about a potential refactor, this is non-blocking, but I'd be curious to see if it makes the logic read easier -- the intermingling of single/multi chunk behavior detracts from what to me looks like a great refactor of the library.

Overall, I really like your implementation, feels like a solid cleanup to the library, and you've done a great job of adding tests.

The only blocker for me, would be I'd like to advocate for the major bump to this library -- curious what others think.

d-goog committed:
fix: Fix generic compatibility

tritone commented on README.md:
```diff
@@ -292,7 +292,19 @@ The maximum number of retries to attempt. Default: 5.
+
+Enables [Multiple chunk upload](https://cloud.google.com/storage/docs/performing-resumable-uploads#chunked-upload) mode and sets each request size to this amount.
+
+This only makes sense to use for larger files. The chunk size should be a multiple of 256 KiB (256 x 1024 bytes), unless it's the last chunk that completes the upload. Larger chunk sizes typically make uploads more efficient. We recommend using at least 8 MiB for the chunk size.
```
The comment about the last chunk is not really relevant here since they are setting this param once before the upload is initialized.

danielbankhead replied:
Thanks, updated

d-goog committed:
docs: Remove note on 'last chunk'

d-goog committed:
fix: Fix type error from generic

ddelgrosso1 commented:
> The only blocker for me, would be I'd like to advocate for the major bump to this library -- curious what others think.

I am onboard with making this a major version bump. I think this library doesn't see much usage as is. I don't think we will have to be porting many fixes backwards, which as I learned can be a painful exercise.

That said, if others object I want to just triple check that the API surface wasn't changed as a result of this refactor?

danielbankhead commented:
Thanks for the feedback @bcoe, @ddelgrosso1, & @tritone - labeled this as breaking.
